packages/components/src/higher-order/with-fallback-styles/index.js

```diff
@@ -38,9 +40,24 @@ export default ( mapNodeToProps ) => createHigherOrderComponent(
 				this.grabFallbackStyles();
 			}
 
+			shouldRecomputeValues() {
+				let lastRecomputeValues = [];
+				return () => {
+					if ( ! computeWhenChange ) {
+						return ! this.state.grabStylesCompleted;
+					}
+					const newRecomputeValues = computeWhenChange( this.props );
+					if ( isShallowEqual( lastRecomputeValues, newRecomputeValues ) ) {
+						return false;
+					}
+					lastRecomputeValues = newRecomputeValues;
+					return true;
+				};
+			}
+
 			grabFallbackStyles() {
-				const { grabStylesCompleted, fallbackStyles } = this.state;
-				if ( this.nodeRef && ! grabStylesCompleted ) {
+				const { fallbackStyles } = this.state;
+				if ( this.nodeRef && this.shouldRecomputeValues() ) {
```

Comment:
Is `mapNodeToProps` that expensive to run that we need to cache it?

It looks like it would solve the issue if we would avoid the local state in the first place. Otherwise, we need to find a way to reset it when the styles change outside, e.g. you apply a class name to the block or change its style variant. which is not taken into account in the proposed solution.